designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,808 @@
+* `defs` - Contains definitions that are used in the config.
+  * `ruleNamespaces` - Contains definitions for rule namespaces grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `globals`
+* `settings`
+* `rules`
+* `parserOptions`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `parser` - an object in `eslint.config.js` files (a string in `.eslintrc`)
+* `processor` - an object in `eslint.config.js` files (a string in `.eslintrc`)
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `eslint.config.js`, these are all objects, requiring users to manually specify the objects to use.
+
+The following keys are invalid in `eslint.config.js`:
+
+* `extends` - replaced by config arrays
+* `env` - responsibility of the user
+* `overrides` - responsibility of the user
+* `plugins` - replaced by `ruledefs`
+* `root` - always considered `true`
+
+Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.
+
+#### Extending Another Config
+
+Extending another config is accomplished by returning an array as the value of `exports.config`. Configs that come later in the array are merged with configs that come earlier in the array. For example:
+
+```js
+exports.config = [
+    require("eslint-config-standard"),
+    {
+        files: ["*.js"],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+This config extends `eslint-config-standard` because that package is included first in the array. You can add multiple configs into the array to extend from multiple configs, such as:
+
+```js
+exports.config = [
+    require("eslint-config-standard"),
+    require("@me/eslint-config"),
+    {
+        files: ["*.js"],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+Each item in a config array can be a config array. For example, this is a valid config array and equivalent to the previous example:
+
+```js
+exports.config = [
+    [
+        require("eslint-config-standard"),
+        require("@me/eslint-config")
+    ],
+    {
+        files: ["*.js"],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+A config array is always flattened before being evaluated, so even though this example is a two-dimensional config array, it will be evaluated as if it were a one-dimensional config array.
+
+When using a config array, only one config object must have a `files` key (config arrays where no objects contain `files` will result in an error). If a config in the config array does not contain `files` or `ignores`, then that config is merged into every config with a `files` pattern. For example:
```

Comment:
I wonder if `ignores` in shared configs or plugins should be allowed. My concern is that a plugin or shared config's `ignores` could be updated and the linting task could begin silently ignoring files the user is expecting to be linted.

Reply:
I've changed my mind on this, as I think the benefits of having a shared `ignores` key in a base config outweighs the potential downside of things silently being ignored (though this is still a real danger).